### ISSUE TITLE
[inductor][cpu]GPT2ForSequenceClassification AMP static/dynamic shape default/cpp wrapper single thread accuracy crash

### ISSUE THREAD

**Issue by zxd1997066** [pytorch/pytorch#123503]
### 🐛 Describe the bug

```
loading model: 0it [00:02, ?it/s]cpu  eval  GPT2ForSequenceClassification      

E0402 16:13:08.429653 134289029980032 torch/_dynamo/utils.py:1405] RMSE (res-fp64): 0.00568, (ref-fp64): 0.00100 and shape=torch.Size([1, 2]). res.dtype: torch.bfloat16, multiplier: 3.000000, tol: 0.004000
E0402 16:13:08.429873 134289029980032 torch/_dynamo/utils.py:1319] Accuracy failed for key name logits
fail_accuracy
```

### Versions

</table><p>SW info</p><table border="1" class="dataframe table">
  <thead>
    <tr style="text-align: right;">
      <th>name</th>
      <th>target_branch</th>
      <th>target_commit</th>
      <th>refer_branch</th>
      <th>refer_commit</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>torchbench</td>
      <td>main</td>
      <td>d6015d42</td>
      <td>main</td>
      <td>d6015d42</td>
    </tr>
    <tr>
      <td>torch</td>
      <td>main</td>
      <td>781e8d2201c1e2aaeccbbc7b7b13f9322b481bc9</td>
      <td>main</td>
      <td>f0d461beacded34abe196c72ec4bcdb55bf01793</td>
    </tr>
    <tr>
      <td>torchvision</td>
      <td>main</td>
      <td>0.19.0a0+2c4665f</td>
      <td>main</td>
      <td>0.19.0a0+a0c79b3</td>
    </tr>
    <tr>
      <td>torchtext</td>
      <td>main</td>
      <td>0.16.0a0+b0ebddc</td>
      <td>main</td>
      <td>0.16.0a0+b0ebddc</td>
    </tr>
    <tr>
      <td>torchaudio</td>
      <td>main</td>
      <td>2.2.0a0+ea437b3</td>
      <td>main</td>
      <td>2.2.0a0+17a7081</td>
    </tr>
    <tr>
      <td>torchdata</td>
      <td>main</td>
      <td>0.7.1a0+0790338</td>
      <td>main</td>
      <td>0.7.1a0+0790338</td>
    </tr>
    <tr>
      <td>dynamo_benchmarks</td>
      <td>main</td>
      <td>nightly</td>
      <td>main</td>
      <td>nightly</td>
    </tr>
  </tbody>
</table>

</table>

Repro:
[inductor_single_run.sh](https://github.com/chuanqi129/inductor-tools/blob/main/scripts/modelbench/inductor_single_run.sh)
bash inductor_single_run.sh single inference accuracy huggingface GPT2ForSequenceClassification amp first static/dynamic default/cpp
[huggingface-GPT2ForSequenceClassification-inference-amp-static-cpp-single-accuracy-crash_guilty_commit.log](https://github.com/pytorch/pytorch/files/14893899/huggingface-GPT2ForSequenceClassification-inference-amp-static-cpp-single-accuracy-crash_guilty_commit.log)
Suspected guilty commit: https://github.com/pytorch/pytorch/commit/5152945441b2a1387f6e9fc8fe1531bfeb6cabe4
cc @ezyang @msaroufim @bdhirsh @anijain2305 @chauhang @WeizhuoZhang-intel @chuanqi129

**Comment by leslie-fang-intel** [comment]
@sanchitintel could you help to take a look of this issue?

**Comment by sanchitintel** [comment]
Sorry, had missed the notification. Weird that only the single-threaded case has an accuracy issue. Will investigate, thanks!

**Comment by leslie-fang-intel** [comment]
Hi @sanchitintel, any update for this issue?

**Comment by leslie-fang-intel** [comment]
Hi @sanchitintel, have you got time to take a look of this issue?

**Comment by sanchitintel** [comment]
Sorry, I was unable to find the root-cause of this one, since the accuracy issue only happens with single-threaded case.
As we discussed offline, I'll add a `check_fn` to disable this SDPA pattern for the single-thread case on CPU.

Thanks

**Comment by sanchitintel** [comment]
Hi @Valentine233 @leslie-fang-intel, this issue is happening with AMP with BF16 dtype because of MKL's `GEMM_BF16BF16F32` kernels when `AMX_BF16` ISA is used. The MKL version I used locally is 2024.1.0.

It's even reproducible with multiple threads on the machine with a Xeon 4th generation processor I'm currently using with the PyTorch main branch, although on the machines @zxd1997066 is using, it's not reproducible with multiple threads:

```
python benchmarks/dynamo/huggingface.py --accuracy --amp -dcpu -n50 --no-skip --dashboard --batch-size 1 --threads 1  --timeout 9000 --backend=inductor --inference --only GPT2ForSequenceClassification
```
When AMX is used (default case), this issue happens.
With environment variable[ `MKL_ENABLE_INSTRUCTIONS=AVX512_E3`](https://www.intel.com/content/www/us/en/docs/onemkl/developer-guide-windows/2023-0/instruction-set-specific-dispatch-on-intel-archs.html) is used, this issue doesn't occur. `AVX512_E3` here means AVX512_BF16 ISA but not AMX_BF16.

Looks like accuracy should be checked for these MKL kernels for those input shapes with `AMX_BF16` ISA. .
Will contact our MKL team, and will try to have this pattern disabled on machines with `AMX_BF16` ISA.
However, that requires AMX detection in python, and it's being enabled as a part of #127195, so could we temporarily disable this pattern for every ISA, and request such a PR to be cherry-picked to ? 

Thanks!

**Comment by sanchitintel** [comment]
After also varying the maximum ISA oneDNN could use, it's now unclear as to whether MKL `GEMM_BF16BF16F32`s are actually responsible.

| MKL also uses AMX_BF16 besides AVX512_BF16| MKL only uses AVX512_BF16 | oneDNN also uses AMX_BF16 besides AVX512_BF16 | oneDNN only uses AVX512_BF16 | Does accuracy failure occur (with 1 to 16 threads)? |
|--------|----------------|--------------|---------------------|--------------------|
|Y| N             | Y                   | N               |Y|              
|Y| N | N | Y | N|
|N| Y | N | Y | N|
|N| Y | Y | N | N|

Example command:
```
DNNL_MAX_CPU_ISA=AVX512_CORE_BF16 ONEDNN_VERBOSE=1 MKL_ENABLE_INSTRUCTIONS=AVX512_E4  MKL_VERBOSE=1 OMP_NUM_THREADS=8 MKL_NUM_THREADS=8 TORCHINDUCTOR_FREEZING=0 python benchmarks/dynamo/huggingface.py --accuracy --amp -dcpu -n50 --no-skip --dashboard --batch-size 1 --threads 8  --timeout 9000 --backend=inductor --inference --only GPT2ForSequenceClassification
```

**Next action item:** Since the commit responsible for this issue enabled offloading of SDPA compute to the Flash Attention kernel, a BF16 AMP UT would make things clear because currently, no CI machines test for AMP variant of SDPA patterns with AMX ISA.

**Comment by leslie-fang-intel** [comment]
> After also varying the maximum ISA oneDNN could use, it's now unclear as to whether MKL `GEMM_BF16BF16F32`s are actually responsible.
> 
> MKL could also use AMX_BF16 besides AVX512_BF16	MKL only used AVX512_BF16	oneDNN could also use AMX_BF16 besides AVX512_BF16	oneDNN only used AVX512_BF16	Does accuracy failure occur (with 1 to 16 threads)?
> Y	N	Y	N	Y
> Y	N	N	Y	N
> N	Y	N	Y	N
> N	Y	Y	N	N
> Example command:
> 
> ```
> DNNL_MAX_CPU_ISA=AVX512_CORE_BF16 ONEDNN_VERBOSE=1 MKL_ENABLE_INSTRUCTIONS=AVX512_E4  MKL_VERBOSE=1 OMP_NUM_THREADS=8 MKL_NUM_THREADS=8 TORCHINDUCTOR_FREEZING=0 python benchmarks/dynamo/huggingface.py --accuracy --amp -dcpu -n50 --no-skip --dashboard --batch-size 1 --threads 8  --timeout 9000 --backend=inductor --inference --only GPT2ForSequenceClassification
> ```
> 
> **Next action item:** Since the commit responsible for this issue enabled offloading of SDPA compute to the Flash Attention kernel, a BF16 AMP UT would make things clear because currently, no CI machines with AMX ISA test for AMP variants of SDPA patterns.

Thanks for the investigation. Are these experiments all with SDPA fusion pattern 18 enabled? Comparing the the first and second experiments, it looks like something from oneDNN side causes the discrepancy which seems not aligned with the guilty commit found in this issue summary.

**Comment by sanchitintel** [comment]
> Comparing the the first and second experiments, it looks like something from oneDNN side causes the discrepancy which seems not aligned with the guilty commit found in this issue summary

Hi @leslie-fang-intel, the 3rd & 4th experiments rule out the possibility of oneDNN being an issue.
It seems this issue wasn't caught with BF16 because the CI machines don't support AMX.
If I add BF16 support to SDPA pattern UTs (which only test for FP32 at this point), and test on a machine that supports AVX512_BF16 or AMX, the UTs for patterns 18 & 19 fail for BF16 dtype because of exceeding tolerance bounds.

e.g.
```
Mismatched elements: 203 / 1024 (19.8%)
Greatest absolute difference: 0.015625 at index (0, 1, 1, 27) (up to 1e-05 allowed)
Greatest relative difference: 1.0859375 at index (0, 15, 1, 16) (up to 1.3e-06 allowed)
```

cc @Valentine233

**Comment by sanchitintel** [comment]
This issue still persists for BF16 dtype with Inductor SDPA pattern 18 that I added, and pattern 19 that @Valentine233 added.

**Comment by leslie-fang-intel** [comment]
Running with Max-autotune, the model passes accuracy check
```
TORCHINDUCTOR_MAX_AUTOTUNE_GEMM_BACKENDS="CPP" TORCHINDUCTOR_MAX_AUTOTUNE=1 numactl -C 56-56 --membind=1 python benchmarks/dynamo/huggingface.py --accuracy --amp -dcpu -n50 --no-skip --dashboard --batch-size 1 --threads 1 --only GPT2ForSequenceClassification --inference --freezing --timeout 9000 --backend=inductor --output=/tmp/inductor_single_test_st.csv
```

One difference found w and wo max-autotune is the post op of `cpp_fused_add_mul_pow_tanh_27`. With max-autotune, these post ops fused in gemm template and directly calculated with fp32 without cvt to bf16 (mkldnn linear out) and cvt to fp32 (legalization of cpp codegen). cc @Valentine233.

**Comment by chunyuan-w** [comment]
AOTI also has the same issue caused by the same guilty commit.

**Comment by sanchitintel** [comment]
> AOTI also has the same issue caused by the same guilty commit.

Also with a single thread?

**Comment by chunyuan-w** [comment]
> > AOTI also has the same issue caused by the same guilty commit.
> 
> Also with a single thread?

Both single and multiple thread have accuracy failures for GPT2ForSequenceClassification in AOTI. @zxd1997066 could you help confirm if both single and multiple thread failures in AOTI are caused by the guilty commit in this issue?

**Comment by Valentine233** [comment]
@zxd1997066 Please help check if the issue is fixed, thanks!
